Proposals/NNNN-swift-subprocess.md

```diff
@@ -0,0 +1,882 @@
+            launchRequirementData: Data? = nil,
+            additionalSpawnAttributeConfiguration: (@Sendable (inout posix_spawnattr_t?) throws -> Void)? = nil,
+            additionalFileAttributeConfiguration: (@Sendable (inout posix_spawn_file_actions_t?) throws -> Void)? = nil
+        )
+
+        public static var `default`: Self
+    }
+}
+```
+
+`PlatformOptions` also supports "escape hatches" that enable developers to configure the underlying platform-specific objects directly if `Subprocess` lacks corresponding high-level APIs.
+
+For Darwin, we are proposing two such APIs:
+
+- `.additionalSpawnAttributeConfiguration: (@Sendable (inout posix_spawnattr_t?) throws -> Void)?` gives developers an opportunity to configure the `posix_spawnattr_t` object just before it's passed to `posix_spawn()`. For instance, developers can set additional spawn flags:
+
+```swift
+let config = Subprocess.Configuration(executing: .at("/my/executable"))
+config.additionalSpawnAttributeConfiguration = { spawnAttr in
+    let flags: Int32 = POSIX_SPAWN_CLOEXEC_DEFAULT |
+        POSIX_SPAWN_SETSIGMASK |
+        POSIX_SPAWN_SETSIGDEF |
+        POSIX_SPAWN_START_SUSPENDED
+    posix_spawnattr_setflags(&spawnAttr, Int16(flags))
+}
+```
+
+- Similarly, `.additionalFileAttributeConfiguration: (@Sendable (inout posix_spawn_file_actions_t?) throws -> Void)` allows developers to customize `posix_spawn_file_actions_t`. For instance, a developer might want to bind child file descriptors, other than standard input (fd 0), standard output (fd 1), and standard error (fd 2), to parent file descriptors:
+
+```swift
+let config = Subprocess.Configuration(executing: .at("/my/executable"))
+// Bind child fd 4 to a parent fd
+config.additionalFileAttributeConfiguration = { fileAttr in
+    let parentFd: FileDescriptor = ...
+    posix_spawn_file_actions_adddup2(&fileAttr, parentFd.rawValue, 4)
+} 
+```
+
+_(We welcome community input on which Linux and Windows "escape hatches" we should add)_
+
+
+### `Subprocess.InputMethod`
+
+In addition to supporting the direct passing of `Sequence<UInt8>` and `AsyncSequence<UInt8>` as the standard input to the child process, `Subprocess` also provides a `Subprocess.InputMethod` type that includes two additional input options:
+- `.noInput`: Specifies that the subprocess does not require any standard input. This is the default value.
+- `.readFrom`: Specifies that the subprocess should read its standard input from a file descriptor provided by the developer. Subprocess will automatically close the file descriptor after the process exits if `closeWhenDone` is set to `true`.
+
+```swift
+extension Subprocess {
+    @available(FoundationPreview 0.4, *)
+    @available(iOS, unavailable)
+    @available(tvOS, unavailable)
+    @available(watchOS, unavailable)
+    public struct InputMethod: Sendable, Hashable {
+        public static var noInput: Self
+        public static func readFrom(_ fd: FileDescriptor, closeWhenDone: Bool) -> Self
```

Comment:
It's crucial here to specify who closes this file descriptor in the parent and precisely when and how. Please note that these file descriptors must usually be closed in the parent _right after_ fork/posix_spawn. It is _not_ acceptable to wait with the closing until after the process exits because that may lead to a deadlock. For example, the parent commonly inherits a pipe into the child. But the child can only see EOF on the pipe if there are no more writers left. Also: It's crucial to keep the number of open file descriptors low (macOS default limit is 256).

I would assume that `Subprocess` either always closes or has the ability to close file descriptors that it inherits into a child at the earliest possible point. But there doesn't seem to be API or docs around this (unless I missed it).

Reply:
(same for `writeTo`).